core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java

```diff
@@ -217,6 +217,28 @@ public void stop() {
       localStage.whenComplete(biConsumer);
    }
 
+   private <I, R> void handleNoTargets(Function<? super Publisher<I>, ? extends CompletionStage<R>> transformer,
+         Function<? super Publisher<R>, ? extends CompletionStage<R>> finalizer,
+         FlowableProcessor<R> flowableProcessor) {
+      // Need to do this in case if the transformer or finalizer has additional values such as reduce with identity
+      // or switchIfEmpty etc.
+      CompletionStage<R> transformedStage = transformer.apply(Flowable.empty());
+      if (finalizer != null) {
+         transformedStage = transformedStage.thenApply(value -> finalizer.apply(Flowable.just(value)))
+               .thenCompose(Function.identity());
+      }
```

Comment:
Another issue is that `transformedStage` could complete with `null` and `Flowable.just` would then throw an NPE. This happens when `DistributedCacheStream.reduce(BinaryOperator<R>)` (specifies no identity element) is used.

I crafted this unit-test to show the point:
```
   @Test(dataProvider = "GuaranteeParallelEntry")
   public void testReduceWithoutIdentityEmptyKeys(DeliveryGuarantee deliveryGuarantee, boolean parallel, boolean isEntry) {
      Cache<Integer, String> cache = cache(0);
      int insertAmount = insert(cache).size();
      assertTrue(insertAmount > 0);

      // This empty set filters everything, but the computation must still complete
      Set<Integer> keysToInclude = new HashSet<>();

      // Using a reduce function without explicit identity tests the special case where completion-stages terminate with null
      ClusterPublisherManager<Integer, String> cpm = cpm(cache);
      CompletionStage<?> stage;
      if (isEntry) {
         Function<Publisher<CacheEntry<Integer, String>>, CompletionStage<CacheEntry<Integer, String>>> function =
               PublisherReducers.reduce((SerializableBinaryOperator<CacheEntry<Integer, String>>) (e1, e2) -> e1); // reduce without identity
         stage = cpm.entryReduction(parallel, null, keysToInclude, null, false, deliveryGuarantee,
               function, function);
      } else {
         Function<Publisher<Integer>, CompletionStage<Integer>> function =
               PublisherReducers.reduce((SerializableBinaryOperator<Integer>) (k1, k2) -> k1); // reduce without identity
         stage = cpm.keyReduction(parallel, null, keysToInclude, null, false, deliveryGuarantee,
               function, function);
      }

      Object result = stage.toCompletableFuture().join();
      assertTrue(result == null);
   }
```

**NOTE:** This unit-test comes from another (really small) PR I was preparing, so it would also fail with my original PR code. Wanted to see how this PR goes till I submit the next ;-)

Reply:
This is fixed by the previous comment, just removing the finalizer as the whenComplete properly handles null. But I can add this test as well.